File: src/lib/api/investments.ts
import { apiGet, apiPost, apiPut } from "$lib/api/client";

function buildQuery(params: Record<string, string | number | null | undefined>): string {
  const searchParams = new URLSearchParams();

  for (const [key, value] of Object.entries(params)) {
    if (value !== null && value !== undefined) {
      searchParams.set(key, String(value));
    }
  }

  const query = searchParams.toString();
  return query ? `?${query}` : "";
}

export type InvestmentInstrumentCreateInput = {
  book_id: number;
  commodity_id: number;
  instrument_type: string;
  display_name: string;
  symbol: string | null;
  isin: string | null;
  cusip: string | null;
  figi: string | null;
  exchange: string | null;
  venue: string | null;
  issuer: string | null;
  country_code: string | null;
  quote_commodity_id: number | null;
  trading_commodity_id: number | null;
  quantity_scale: number;
  price_scale: number;
  is_active: boolean;
  metadata_json: string | null;
};

export type InvestmentInstrumentUpdateInput = InvestmentInstrumentCreateInput & { id: number };

export type InvestmentInstrumentSummary = InvestmentInstrumentCreateInput & {
  id: number;
  created_at: string;
  updated_at: string;
};

export type CostBasisProfileCreateInput = {
  book_id: number;
  name: string;
  method: string;
  description: string | null;
  is_default: boolean;
  metadata_json: string | null;
};

export type CostBasisProfileUpdateInput = CostBasisProfileCreateInput & { id: number };

export type CostBasisProfileSummary = CostBasisProfileCreateInput & {
  id: number;
  created_at: string;
  updated_at: string;
};

export type CorporateActionCreateInput = {
  book_id: number;
  action_type: string;
  old_instrument_id: number | null;
  new_instrument_id: number | null;
  effective_date: string;
  ratio_numerator: number | null;
  ratio_denominator: number | null;
  cash_in_lieu_minor: number | null;
  cash_commodity_id: number | null;
  source_reference: string | null;
  memo: string | null;
  metadata_json: string | null;
};

export type CorporateActionSummary = CorporateActionCreateInput & {
  id: number;
  generated_transaction_id: number | null;
  created_at: string;
};

export type BuyCommodityInput = {
  book_id: number;
  txn_date: string;
  commodity_id: number;
  investment_account_id: number;
  cash_account_id: number;
  quantity_minor: number;
  cash_amount_minor: number;
  memo: string | null;
  payee_id: number | null;
  status: string | null;
};

export type SellLotAllocationInput = {
  lot_id: number;
  quantity_minor: number;
};

export type SellCommodityInput = {
  book_id: number;
  txn_date: string;
  commodity_id: number;
  investment_account_id: number;
  cash_account_id: number;
  quantity_minor: number;
  cash_amount_minor: number;
  lot_strategy: string;
  lot_allocations: SellLotAllocationInput[] | null;
  allow_short: boolean;
  memo: string | null;
  payee_id: number | null;
  status: string | null;
};

export type DividendInput = {
  book_id: number;
  txn_date: string;
  cash_account_id: number;
  income_account_id: number;
  amount_minor: number;
  memo: string | null;
  payee_id: number | null;
  status: string | null;
};

export type ReinvestedDividendInput = {
  book_id: number;
  txn_date: string;
  commodity_id: number;
  investment_account_id: number;
  income_account_id: number;
  quantity_minor: number;
  amount_minor: number;
  memo: string | null;
  payee_id: number | null;
  status: string | null;
};

export type TradeAllocation = {
  lot_id: number;
  quantity_minor: number;
};

export type TradeResult = {
  transaction_id: number;
  allocations: TradeAllocation[];
  lot_id: number | null;
};

export type DividendResult = {
  transaction_id: number;
};

export type InvestmentPositionLot = {
  lot_id: number;
  opened_date: string | null;
  quantity_minor: number;
  cost_basis_minor: number;
  remaining_cost_basis_minor: number;
  converted_value_minor: number | null;
  converted_cost_basis_minor: number | null;
  price_missing: boolean | null;
};

export type InvestmentPositionSummary = {
  account_id: number;
  account_name: string;
  account_type: string;
  commodity_id: number;
  commodity_name: string;
  commodity_scale: number;
  balance_minor: number;
  lots: InvestmentPositionLot[];
};

export type ConvertedInvestmentPositionSummary = InvestmentPositionSummary & {
  value_minor: number;
  price_missing: boolean;
};

export type InvestmentLotHoldingSummary = {
  lot_id: number;
  account_id: number;
  account_name: string;
  commodity_id: number;
  commodity_name: string;
  opened_date: string | null;
  quantity_minor: number;
  cost_basis_minor: number;
  commodity_scale: number;
  holding_days: number | null;
  is_long_term: boolean;
};

export type PortfolioPerformanceSummary = {
  book_id: number;
  base_commodity_id: number;
  as_of_date: string | null;
  market_value_minor: number;
  cost_basis_minor: number;
  unrealized_gain_minor: number;
  realized_gain_minor: number;
  income_minor: number;
  price_missing_count: number;
};

export type AccountValuationRow = {
  account_id: number;
  account_name: string;
  account_commodity_id: number;
  market_value_minor: number;
  cost_basis_minor: number;
  unrealized_gain_minor: number;
  price_missing: boolean;
  fx_missing: boolean;
};

export type CurrencyExposureRow = {
  commodity_id: number;
  commodity_name: string;
  quantity_minor: number;
  native_value_minor: number;
  price_missing: boolean;
};

export async function getPositions(bookId: number, asOfDate?: string | null): Promise<InvestmentPositionSummary[]> {
  const path = `/investments/positions${buildQuery({ book_id: bookId, as_of_date: asOfDate ?? null })}`;
  return apiGet<InvestmentPositionSummary[]>(path);
}

export async function convertPositions(
  bookId: number,
  baseCommodityId: number,
  asOfDate?: string | null
): Promise<ConvertedInvestmentPositionSummary[]> {
  const path = `/investments/positions/converted${buildQuery({
    book_id: bookId,
    base_commodity_id: baseCommodityId,
    as_of_date: asOfDate ?? null,
  })}`;
  return apiGet<ConvertedInvestmentPositionSummary[]>(path);
}

export async function listLotsWithHoldingPeriod(
  bookId: number,
  accountId?: number | null,
  commodityId?: number | null,
  asOfDate?: string | null
): Promise<InvestmentLotHoldingSummary[]> {
  const path = `/investments/lots${buildQuery({
    book_id: bookId,
    account_id: accountId ?? null,
    commodity_id: commodityId ?? null,
    as_of_date: asOfDate ?? null,
  })}`;
  return apiGet<InvestmentLotHoldingSummary[]>(path);
}

export async function listInvestmentInstruments(bookId: number): Promise<InvestmentInstrumentSummary[]> {
  return apiGet<InvestmentInstrumentSummary[]>(`/investments/instruments?book_id=${bookId}`);
}

export async function saveInvestmentInstrument(
  input: InvestmentInstrumentCreateInput | InvestmentInstrumentUpdateInput
): Promise<InvestmentInstrumentSummary> {
  if ("id" in input) {
    return apiPut<InvestmentInstrumentSummary, InvestmentInstrumentCreateInput>(`/investments/instruments/${input.id}`, input);
  }
  return apiPost<InvestmentInstrumentSummary, InvestmentInstrumentCreateInput>("/investments/instruments", input);
}

export async function listCostBasisProfiles(bookId: number): Promise<CostBasisProfileSummary[]> {
  return apiGet<CostBasisProfileSummary[]>(`/investments/cost-basis-profiles?book_id=${bookId}`);
}

export async function saveCostBasisProfile(
  input: CostBasisProfileCreateInput | CostBasisProfileUpdateInput
): Promise<CostBasisProfileSummary> {
  if ("id" in input) {
    return apiPut<CostBasisProfileSummary, CostBasisProfileCreateInput>(`/investments/cost-basis-profiles/${input.id}`, input);
  }
  return apiPost<CostBasisProfileSummary, CostBasisProfileCreateInput>("/investments/cost-basis-profiles", input);
}

export async function listCorporateActions(bookId: number): Promise<CorporateActionSummary[]> {
  return apiGet<CorporateActionSummary[]>(`/investments/corporate-actions?book_id=${bookId}`);
}

export async function createCorporateAction(input: CorporateActionCreateInput): Promise<CorporateActionSummary> {
  return apiPost<CorporateActionSummary, CorporateActionCreateInput>("/investments/corporate-actions", input);
}

export async function buyCommodity(input: BuyCommodityInput): Promise<TradeResult> {
  return apiPost<TradeResult, BuyCommodityInput>("/investments/buy", input);
}

export async function sellCommodity(input: SellCommodityInput): Promise<TradeResult> {
  return apiPost<TradeResult, SellCommodityInput>("/investments/sell", input);
}

export async function createDividendTransaction(input: DividendInput): Promise<DividendResult> {
  return apiPost<DividendResult, DividendInput>("/investments/dividend", input);
}

export async function createReinvestedDividend(input: ReinvestedDividendInput): Promise<TradeResult> {
  return apiPost<TradeResult, ReinvestedDividendInput>("/investments/reinvested-dividend", input);
}

export async function getPortfolioPerformance(
  bookId: number,
  baseCommodityId: number,
  asOfDate?: string | null,
  costBasisProfileId?: number | null
): Promise<PortfolioPerformanceSummary> {
  const path = `/investments/performance${buildQuery({
    book_id: bookId,
    base_commodity_id: baseCommodityId,
    as_of_date: asOfDate ?? null,
    cost_basis_profile_id: costBasisProfileId ?? null,
  })}`;
  return apiGet<PortfolioPerformanceSummary>(path);
}

export async function getAccountValuation(
  bookId: number,
  baseCommodityId: number,
  asOfDate?: string | null,
  costBasisProfileId?: number | null
): Promise<AccountValuationRow[]> {
  const path = `/investments/account-valuation${buildQuery({
    book_id: bookId,
    base_commodity_id: baseCommodityId,
    as_of_date: asOfDate ?? null,
    cost_basis_profile_id: costBasisProfileId ?? null,
  })}`;
  return apiGet<AccountValuationRow[]>(path);
}

export async function getCurrencyExposure(bookId: number, asOfDate?: string | null): Promise<CurrencyExposureRow[]> {
  const path = `/investments/currency-exposure${buildQuery({ book_id: bookId, as_of_date: asOfDate ?? null })}`;
  return apiGet<CurrencyExposureRow[]>(path);
}

File: apps/api/src/rekenraam_api/schemas/investments.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class PositionsQuery(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int
    as_of_date: date | None = None


class ConvertedPositionsQuery(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int
    base_commodity_id: int
    as_of_date: date | None = None


class LotsHoldingQuery(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int
    account_id: int | None = None
    commodity_id: int | None = None
    as_of_date: date | None = None


class RealizedGainsQuery(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int
    date_from: date | None = None
    date_to: date | None = None
    cost_basis_profile_id: int | None = None


class UnrealizedGainsQuery(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int
    base_commodity_id: int
    as_of_date: date | None = None
    cost_basis_profile_id: int | None = None


class InvestmentInstrumentCreateInput(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int = 1
    commodity_id: int
    instrument_type: str
    display_name: str
    symbol: str | None = None
    isin: str | None = None
    cusip: str | None = None
    figi: str | None = None
    exchange: str | None = None
    venue: str | None = None
    issuer: str | None = None
    country_code: str | None = None
    quote_commodity_id: int | None = None
    trading_commodity_id: int | None = None
    quantity_scale: int = 4
    price_scale: int = 4
    is_active: bool = True
    metadata_json: str | None = None


class InvestmentInstrumentUpdateInput(InvestmentInstrumentCreateInput):
    model_config = ConfigDict(frozen=True)


class InvestmentInstrumentSummary(InvestmentInstrumentCreateInput):
    model_config = ConfigDict(frozen=True)

    id: int
    created_at: datetime
    updated_at: datetime


class CostBasisProfileCreateInput(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int = 1
    name: str
    method: str = "fifo"
    description: str | None = None
    is_default: bool = False
    metadata_json: str | None = None


class CostBasisProfileUpdateInput(CostBasisProfileCreateInput):
    model_config = ConfigDict(frozen=True)


class CostBasisProfileSummary(CostBasisProfileCreateInput):
    model_config = ConfigDict(frozen=True)

    id: int
    created_at: datetime
    updated_at: datetime


class CorporateActionCreateInput(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int = 1
    action_type: str
    old_instrument_id: int | None = None
    new_instrument_id: int | None = None
    effective_date: date
    ratio_numerator: int | None = None
    ratio_denominator: int | None = None
    cash_in_lieu_minor: int | None = None
    cash_commodity_id: int | None = None
    source_reference: str | None = None
    memo: str | None = None
    metadata_json: str | None = None


class CorporateActionSummary(CorporateActionCreateInput):
    model_config = ConfigDict(frozen=True)

    id: int
    generated_transaction_id: int | None = None
    created_at: datetime


class BuyCommodityInput(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int = 1
    txn_date: date
    commodity_id: int
    investment_account_id: int
    cash_account_id: int
    quantity_minor: int
    cash_amount_minor: int
    memo: str | None = None
    payee_id: int | None = None
    status: str | None = None


class SellLotAllocationInput(BaseModel):
    model_config = ConfigDict(frozen=True)

    lot_id: int
    quantity_minor: int


class SellCommodityInput(BaseModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True)

    book_id: int = 1
    txn_date: date
    commodity_id: int
    investment_account_id: int
    cash_account_id: int
    quantity_minor: int
    cash_amount_minor: int
    lot_strategy: str = "default"
    lot_allocations: tuple[SellLotAllocationInput, ...] | None = None
    allow_short: bool = False
    memo: str | None = None
    payee_id: int | None = None
    status: str | None = None


class DividendInput(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int = 1
    txn_date: date
    cash_account_id: int
    income_account_id: int
    amount_minor: int
    memo: str | None = None
    payee_id: int | None = None
    status: str | None = None


class ReinvestedDividendInput(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int = 1
    txn_date: date
    commodity_id: int
    investment_account_id: int
    income_account_id: int
    quantity_minor: int
    amount_minor: int
    memo: str | None = None
    payee_id: int | None = None
    status: str | None = None


class InvestmentEventInput(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int = 1
    action_type: str
    instrument_id: int | None = None
    effective_date: date
    amount_minor: int | None = None
    quantity_minor: int | None = None
    cash_commodity_id: int | None = None
    memo: str | None = None
    metadata_json: str | None = None


class TradeAllocation(BaseModel):
    model_config = ConfigDict(frozen=True)

    lot_id: int
    quantity_minor: int


class TradeResult(BaseModel):
    model_config = ConfigDict(frozen=True)

    transaction_id: int
    allocations: tuple[TradeAllocation, ...]
    lot_id: int | None


class DividendResult(BaseModel):
    model_config = ConfigDict(frozen=True)

    transaction_id: int


class PortfolioPerformanceQuery(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int
    base_commodity_id: int
    as_of_date: date | None = None
    cost_basis_profile_id: int | None = None


class PortfolioPerformanceSummary(BaseModel):
    model_config = ConfigDict(frozen=True)

    book_id: int
    base_commodity_id: int
    as_of_date: date | None
    market_value_minor: int
    cost_basis_minor: int
    unrealized_gain_minor: int
    realized_gain_minor: int
    income_minor: int
    price_missing_count: int


class AccountValuationRow(BaseModel):
    model_config = ConfigDict(frozen=True)

    account_id: int
    account_name: str
    account_commodity_id: int
    market_value_minor: int
    cost_basis_minor: int
    unrealized_gain_minor: int
    price_missing: bool
    fx_missing: bool


class CurrencyExposureRow(BaseModel):
    model_config = ConfigDict(frozen=True)

    commodity_id: int
    commodity_name: str
    quantity_minor: int
    native_value_minor: int
    price_missing: bool


class PositionLot(BaseModel):
    model_config = ConfigDict(frozen=True)

    lot_id: int
    opened_date: date | None
    quantity_minor: int
    cost_basis_minor: int
    remaining_cost_basis_minor: int
    converted_value_minor: int | None = None
    converted_cost_basis_minor: int | None = None
    price_missing: bool | None = None


class Position(BaseModel):
    model_config = ConfigDict(frozen=True)

    account_id: int
    account_name: str
    account_type: str
    commodity_id: int
    commodity_name: str
    commodity_scale: int
    balance_minor: int
    lots: tuple[PositionLot, ...]


class ConvertedPosition(BaseModel):
    model_config = ConfigDict(frozen=True)

    account_id: int
    account_name: str
    account_type: str
    commodity_id: int
    commodity_name: str
    commodity_scale: int
    balance_minor: int
    value_minor: int
    price_missing: bool
    lots: tuple[PositionLot, ...]


class LotHoldingPeriod(BaseModel):
    model_config = ConfigDict(frozen=True)

    lot_id: int
    account_id: int
    account_name: str
    commodity_id: int
    commodity_name: str
    opened_date: date | None
    quantity_minor: int
    cost_basis_minor: int
    commodity_scale: int
    holding_days: int | None
    is_long_term: bool


class RealizedGainEntry(BaseModel):
    model_config = ConfigDict(frozen=True)

    tx_id: int
    txn_date: date
    commodity_id: int
    quantity_minor: int
    proceeds_minor: int
    quote_commodity_id: int | None
    cost_basis_minor: int
    gain_loss_minor: int
    proceeds_missing: bool


class UnrealizedGainEntry(BaseModel):
    model_config = ConfigDict(frozen=True)

    account_id: int
    account_name: str
    account_type: str
    commodity_id: int
    commodity_name: str
    value_minor: int
    cost_basis_minor: int
    unrealized_gain_minor: int
    price_missing: bool
